### PR TITLE
[aten] Move submodules to third_party

### PR DESCRIPTION
**Pull request by orionr**
* Move ATen submodules to root-level third_party
* Update aten_mirror.sh to reflect

Confirmed that PyTorch tests passed.

@ezyang when I run `./tools/aten_mirror.sh` what should `EXTRACTED_REPO` be set to?

Review requested from @ezyang, @zdevito. FYI @Yangqing

### CODE REVIEW

**Comment by ezyang** [comment]
`EXTRACTED_REPO=https://github.com/zdevito/ATen` but it won't actually work for you because you don't have push rights to this repo.

**Review by ezyang — COMMENTED** [review]
Are you sure a submod update on the top level cpuinfo isn't needed? It seems less up-to-date than the one you deleted.

**Comment by orionr** [comment]
I'll defer to @Maratyszcza on whether cpuinfo in `third_party/cpuinfo` should be updated for this or not. My gut is that third_party is set to the right version since he's been managing that?

I've temporarily pushed my latest code to this branch plus the master of my fork to test. However I'm getting a failure on running

```
export EXTRACTED_REPO=https://github.com/orionr/ATen
export FULL_REPO=https://github.com/orionr/pytorch
./tools/aten_mirror.sh
```
(which is slow) of

```
Ref 'refs/heads/temporary-split-branch' was rewritten
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
Auto-merging third_party/tbb
Adding as third_party/tbb~temporary-split-branch instead
Auto-merging third_party/catch
Adding as third_party/catch~temporary-split-branch instead
Removing aten/src/ATen/cpu/cpuinfo
Removing aten/.gitmodules
Automatic merge failed; fix conflicts and then commit the result.
Everything up-to-date
```
When I try to fix conflicts I get
```
git add third_party/catch
error: unable to index file third_party/catch
fatal: updating files failed
```
Investigating.

**Comment by orionr** [comment]
For some reason there was a merge conflict, so
```
git reset fullrepo/master third_party/catch
git reset fullrepo/master third_party/tbb
```
did the trick. Now running 
```
cd aten-export-repo/aten
mkdir build install
cd build
cmake .. -DCMAKE_INSTALL_PREFIX=../install
make install
../tools/run_tests.sh .
cd ..
tools/test_install.sh $(pwd)/install $(pwd)
```

File: aten/src/ATen/CMakeLists.txt
Comment (Maratyszcza) [inline aten/src/ATen/CMakeLists.txt]: I suggest to copy the code in Caffe2 (with modification for `CPUINFO_SOURCE_DIR` w.r.t. ATen path), see: https://github.com/pytorch/pytorch/blob/master/cmake/Dependencies.cmake#L95-L118
This will support some nice options:
- Overriding `cpuinfo` location for development by specifying `-DCPUINFO_SOURCE_DIR=/path/to/your/cpuinfo/checkout`.
- Avoiding double-import of `cpuinfo` when `cpuinfo` is already imported through another dependency.
- Linking `cpuinfo` statically even when other libs are linked dynamically. This prevents issues with mulitple cpuinfo versions overwriting each other, e.g. in conda installation.

**Comment by Maratyszcza** [comment]
Thanks @orionr. It was a bit annoying that `cpuinfo` was downloaded twice on recursive checkout.

**Comment by orionr** [comment]
Okay @ezyang - all local tests passed. Let's wait for CI. Thanks.

File: aten/src/ATen/CMakeLists.txt
Comment (Maratyszcza) [inline aten/src/ATen/CMakeLists.txt]: Need to set `CONFU_DEPENDENCIES_BINARY_DIR` as well, similar to https://github.com/pytorch/pytorch/blob/master/cmake/Dependencies.cmake#L71-L75 (`CONFU_DEPENDENCIES_SOURCE_DIR` is not currently needed as all deps are packaged into submodules).

File: aten/src/ATen/CMakeLists.txt
Comment (Maratyszcza) [inline aten/src/ATen/CMakeLists.txt]: This should be some ATen/PyTorch-specific variable

**Comment by ezyang** [comment]
CC @cpuhrsch who will know what we need for cpuinfo in PyTorch.

File: aten/src/ATen/CMakeLists.txt
Comment (orionr) [reply aten/src/ATen/CMakeLists.txt]: Turns out ATen doesn't specify `/MD` or `/MT` for MSVC, so I'll just set this to `set(CPUINFO_RUNTIME_TYPE "shared" CACHE STRING "")` in all cases. Fix coming soon...

**Comment by cpuhrsch** [comment]
We definitely need the newest version of cpuinfo for various cmake extensions and new features around PowerPC. As long as it's available to the ATen library this should be fine. How can I help with this?

**Comment by Maratyszcza** [comment]
@cpuhrsch It never hurts to bump revision in cpuinfo module. Proper PowerPC support is WIP, IBM is working on it in [github.com/leitao/cpuinfo/tree/ppc64le](https://github.com/leitao/cpuinfo/tree/ppc64le)

**Comment by cpuhrsch** [comment]
@Maratyszcza absolutely agree.

**Comment by orionr** [comment]
Bumping `third_party/cpuinfo` in that last commit. I'll verify tests pass before I merge, but this should be ready for review.

**Comment by orionr** [comment]
Caffe2 conda build was stuck, so canceled. @ezyang, @cpuhrsch, @Maratyszcza , how does this look? Much appreciated.

**Review by ezyang — APPROVED** [review]

**Pull request merged**
